### DISCUSSION TITLE
Change video output resolution

### DISCUSSION THREAD

**Discussion by ghost** [roleoroleo/yi-hack-Allwinner-v2#306] [Ideas]
Is anyone interested in being able to run different output resolutions on their camera(s)? 

I ask because I bought 2 Yi-Iot camera's with nondescript model names of BIG YIIOT PTZ 3MP and YI-3MP-20. One was advertised as 2MP the other 3MP. Both use the same board and 1MP sensor and output 1080p. Technically the sensor is closer to 0.9MP (1280x720). I didn't see the point in running 1080p at the expense of higher transmission bandwidth / storage space and possibly reducing clarity as well so I changed it to it's native resolution of 720p. The rmm app appears to check a config setting for values 1, 2 and 3 being equivalent to 720p, 1080p and 1296p. I didn't find where the config was and instead changed the rmm binary so the 1080p setting was also 720p. Does anyone think it would be nice to have the option to change the config in Yi-Hack? I've noticed at least 2 other camera's posted here running 1296p with a 1080p sensor.

Here's what the difference looks like, left 720p, middle 720p @ 1080p with same camera and right true 1080p using icsee cam. The image was distanced to take up approximately the same sensor area.
![1080](https://user-images.githubusercontent.com/94827997/142975100-50f5b0a8-646a-4b6b-bbf6-51e16a2d84b3.png)

Both Yi-Iot cameras have a device ID (SNo.) beginning with A492D but run different versions of b091qp software.

These are the settings I used IINM

```
b091qp ver 9.0.41.99_202103111513 (512KiB frame buffer)
=======================================================
#define BUF_OFFSET_B091QP 300
#define BUF_SIZE_B091QP 524588			// 512KiB frame buffer
#define FRAME_HEADER_SIZE_B091QP 26
#define DATA_OFFSET_B091QP 4
#define LOWRES_BYTE_B091QP 8
#define HIGHRES_BYTE_B091QP 4

#define RESOLUTION_NONE 0
#define RESOLUTION_LOW  360
#define RESOLUTION_HIGH 720			// 1080? is this important? what about when using 3k?
#define RESOLUTION_BOTH 1080			// 1440?

#define RESOLUTION_HD   720
#define RESOLUTION_FHD  1080
#define RESOLUTION_3K   1296


#define W_LOW 640
#define H_LOW 360
#define W_HD 1280
#define H_HD 720
#define W_FHD 1280				// 1920, being lazy to change code
#define H_FHD 720				// 1080
#define W_3K 2304
#define H_3K 1296


#define SPS_TYPE_B091QP 2			// doesn't matter for now as only one profile, not used 

// probably should use SPS4_2_ but again only one profile  
unsigned char SPS4_1280X720[]       = {0x00, 0x00, 0x00, 0x01, 0x67, 0x64, 0x00, 0x20, 0xAC, 0x2C, 0xA8, 0x05, 0x00, 0x5B, 0x26, 0xE0, 0x20, 0x20, 0x20, 0x40};
unsigned char SPS4_1280X720_TI[]    = {0x00, 0x00, 0x00, 0x01, 0x67, 0x64, 0x00, 0x20, 0xAC, 0x2C, 0xA8, 0x05, 0x00, 0x5B, 0x26, 0xE0, 0x20, 0x20, 0x28, 0x00, 0x00, 0x0F, 0xA0, 0x00, 0x02, 0x71, 0x04, 0x20};  

model_high_res = RESOLUTION_HD;




b091qp ver 9.0.41.99_202104291335 (1MiB frame buffer)
=====================================================
#define BUF_SIZE_B091QP 1048876		// 1MiB Buffer

```
The second firmware uses a bigger frame buffer! otherwise much the same. Perhaps Yi-Hack could report actual sensor image size? Just a thought.

**Comment by roleoroleo** [comment]
Interesting.
So we could have 3 different resolutions: 720p, 1080p and 1296p (and 360).
But I don't know if it's worth the hassle.
The quality of the 720p is better but, IMHO, not so different.
I agree with you: it's not necessary to run 1080p at the expense of higher transmission bandwidth / storage space.
Is there a significant difference between the streams? Or between a jpg snapshots?
Are you able to check it?

**Reply by ghost** [reply]
> Are you able to check it?

I have recorded for 60 seconds a looped 18 second 1080 video @ 30fps displayed on a laptop so not ideal and the camera is a bit too close but hopefully provides somewhere near consistent recording across different resolutions. Camera recording settings were both at 20fps and 1Mbps bitrate.

Original file https://www.pexels.com/video/people-giving-high-fives-1149521/
 The video is of some noisy people moving around giving high 5's.

 | | 720 | 1080 |
 |---|---|---|
 | Yi-Hack SS | 232KB | 375KB | 
 | App SS | 600KB | 1050KB |
 | MP4 SDCard | 13.6MB | 13.9MB |
 | RTSP mkv | 8.4MB | 9.1MB |

Video is limited by the bitrate so similar video stream sizes but can affect the quality, especially when there's a lot of change from frame to frame.

### 720p vlc snapshot (SS)
![vlcsnap-720_1M](https://user-images.githubusercontent.com/94827997/147894875-47a05d98-0c39-4661-b5df-b66eab0733e4.png) 

### 1080p vlc snapshot (SS)
![vlcsnap-1080_1M](https://user-images.githubusercontent.com/94827997/147895034-b1cb6250-743d-4836-ae86-011c2b925bb5.png)

1080p needs the bitrate increased to achieve a similar looking recording
### 1080p @ 2Mbps
![vlcsnap-1080_2M](https://user-images.githubusercontent.com/94827997/147895116-752e0ce5-e97d-42f1-ae17-ce5d6fb4dab6.png)
At 2Mbps files size increases to 21.8MB for the SD Card recording and 16.MB for RTSP. Remember this 1080p video is interpolated (stretched) from the native camera sensor video of 720p so is not any better and possibly worse.

Note the SD Card MP4 has 3 video streams, 1080p/720p @ 20fps, 360 @ 20fps and 360 @ 5fps plus 1 audio stream versus 1 video and 1 audio stream of RTSP. Hence the larger size of the SD Card recording.

So seems an option to change resolution would also benefit an option to change bitrate.

**Comment by ghost** [comment]
Well the width and height sizes can be modified too, the 3 settings are default values I guess and have a configuration somewhere? 

1080p is 2.25x the pixels of 720p so I would expect a significant difference for that, less for 1080p->1296p (1.44x), but haven't tested it. 

My first thought was that people don't realize their sensors are lower resolution than the output, but seems it's just me who is irked by this :) Given that I'm happy to let things be.

**Reply by roleoroleo** [reply]
> My first thought was that people don't realize their sensors are lower resolution than the output, but seems it's just me who is irked by this :)

Nobody knows this cheat.
